File: Arrays_and_Hashing/C++/242_valid_anagram.cpp

/*
* TIME COMPLEXITY - O(nlogn)
* RUNTIME: 9 ms - Beats 55.70% of submissions
* MEMORY - 7.4 MB - Beats 99.91% of submissions
*/

#include <vector>
#include <algorithm>
#include <iostream>
#include <string>
/*
* Thoughtprocess:
* 1. Sort both strings
* 2. Compare the strings
*/

class Solution {
public:
    bool isAnagram(std::string s, std::string t) {
        sort(s.begin(), s.end());
        sort(t.begin(), t.end());
        return s == t;
    }
};

File: Arrays_and_Hashing/Python/242_valid_anagram.py

# RUNTIME: 55 ms, faster than 52.68% of python3 submissions. 
# MEMORY USAGE: 17.5 MB, Beats 6.75% of python3 submissions.

# My thougtprocess:
# 1. Sort both lists
# 2. Compare if the lists are equal

# This could probably be optimized further by using a dictionary to store counts.
# The Counter class from collections module could be used for this purpose.

# Time complexity: O(nlogn)

class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        xs = sorted(s)
        xt = sorted(t)
        return xs == xt